compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/dma_loop_subsumption.mlir

```diff
@@ -0,0 +1,576 @@
+// CHECK:       %[[CIRC_DMA:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       amdaie.controlcode
+// CHECK:         scf.for %[[ARG2:.+]] = %[[C0]] to %[[C6]] step %[[C1]]
+// CHECK:           %[[APPLY:.+]] = affine.apply #[[$MAP]](%[[ARG2]])
+// CHECK:           %[[NPU_DMA:.+]] = amdaie.npu.dma_cpy_nd %[[CIRC_DMA]]([0, 0, %[[APPLY]]] [1, 8, 16] [128, 16, 1], [] [] [])
+// CHECK:           amdaie.npu.dma_wait(%[[NPU_DMA]], S2MM)
+#map = affine_map<(d0) -> (d0 * 16)>
+func.func @npu_dma_cpy_nd_too_many_dims_target_on_l1(%arg0: !amdaie.logicalobjectfifo<memref<1x1x8x16xi32, 2>>, %arg1: !amdaie.logicalobjectfifo<memref<8x16xi32>>) {
+  %c0 = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
+  %c6 = arith.constant 6 : index
+  amdaie.workgroup {
+    %0 = amdaie.circular_dma_cpy_nd(%arg0[] [] [], %arg1[] [] []) : (!amdaie.logicalobjectfifo<memref<1x1x8x16xi32, 2>>, !amdaie.logicalobjectfifo<memref<8x16xi32>>)
+    amdaie.controlcode {
+      scf.for %arg2 = %c0 to %c6 step %c1 {
+        %1 = affine.apply #map(%arg2)
+        %2 = amdaie.npu.dma_cpy_nd %0([0, 0, %1] [1, 8, 16] [128, 16, 1], [] [] [])
+        amdaie.npu.dma_wait(%2, S2MM)
+      }
+      amdaie.end
+    }
+  }
+  return
+}
+
+// -----
+
+// Ensure no modification in case of too many dimensions, i.e. 3 existing
+// dimensions in the case of an `amdaie.npu.dma_cpy_nd` with source on L1.
+// CHECK:       #[[$MAP:.+]] = affine_map<(d0) -> (d0 * 16)>
+// CHECK-LABEL: @npu_dma_cpy_nd_too_many_dims_source_on_l1
+// CHECK-DAG:   %[[C0:.+]] = arith.constant 0 : index
+// CHECK-DAG:   %[[C1:.+]] = arith.constant 1 : index
+// CHECK-DAG:   %[[C6:.+]] = arith.constant 6 : index
+// CHECK:       %[[CIRC_DMA:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       amdaie.controlcode
+// CHECK:         scf.for %[[ARG2:.+]] = %[[C0]] to %[[C6]] step %[[C1]]
+// CHECK:           %[[APPLY:.+]] = affine.apply #[[$MAP]](%[[ARG2]])
+// CHECK:           %[[NPU_DMA:.+]] = amdaie.npu.dma_cpy_nd %[[CIRC_DMA]]([] [] [], [0, 0, %[[APPLY]]] [1, 8, 16] [128, 16, 1])
+// CHECK:           amdaie.npu.dma_wait(%[[NPU_DMA]], S2MM)
+#map = affine_map<(d0) -> (d0 * 16)>
+func.func @npu_dma_cpy_nd_too_many_dims_source_on_l1(%arg0: !amdaie.logicalobjectfifo<memref<1x1x8x16xi32>>, %arg1: !amdaie.logicalobjectfifo<memref<8x16xi32, 2>>) {
+  %c0 = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
+  %c6 = arith.constant 6 : index
+  amdaie.workgroup {
+    %0 = amdaie.circular_dma_cpy_nd(%arg0[] [] [], %arg1[] [] []) : (!amdaie.logicalobjectfifo<memref<1x1x8x16xi32>>, !amdaie.logicalobjectfifo<memref<8x16xi32, 2>>)
+    amdaie.controlcode {
+      scf.for %arg2 = %c0 to %c6 step %c1 {
+        %1 = affine.apply #map(%arg2)
+        %2 = amdaie.npu.dma_cpy_nd %0([] [] [], [0, 0, %1] [1, 8, 16] [128, 16, 1])
+        amdaie.npu.dma_wait(%2, S2MM)
+      }
+      amdaie.end
+    }
+  }
+  return
+}
+
+// -----
+
+// CHECK-LABEL: @npu_dma_cpy_nd_with_for_dependency_on_target
```

Comment:
> There's a circular_dma_cpy_nd operation from L2 to L2 (maybe this should be L3 to L2?)

This pass doesn't make any assumption on the source/target of DMA operations, so L2 -> L2 should work as well. L2 -> L2 can be implemented on HW btw.

> There's a control code (lx6) instruction (aka the dma_cpy_nd op) in a loop, where each iteration of the loop starts a copy of a chunk of shape [1, 8, 16] .
This loop of copies is transformed by the new pass into a single copy, of a tensor of size [6, 1, 8, 16].
Is the idea that back pressure will ensure that the data gets copied in manageable size?

The core optimization is to avoid having to program the shim DMAs over and over as the lx6 is not very fast and also to keep the shim DMA busy for a very long time with a single instruction. 

There might be back-pressure on the L2 side depending on how fast the consumer can write into L2 (while synchronized with other DMAs), but  it might also be the case that the data fabric connection to DDR can't keep up with the shim DMA's requests. Regardless, the shim DMA might work a very long time on this single buffer descriptor instruction.